nixos/modules/services/mail/public-inbox.nix

```diff
@@ -0,0 +1,557 @@
+        '';
+      };
+    };
+    imap = {
+      enable = mkEnableOption "the public-inbox IMAP server";
+      args = mkOption {
+        type = with types; listOf str;
+        default = ["-W0"];
+        description = "Command-line arguments to pass to ${manref "public-inbox-imapd" 1}.";
+      };
+      port = mkOption {
+        type = with types; nullOr port;
+        default = 993;
+        description = ''
+          Listening port.
+          Set to null and use <code>systemd.sockets.public-inbox-imapd.listenStreams</code>
+          if you need a more advanced listening.
+        '';
+      };
+      cert = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "/path/to/fullchain.pem";
+        description = "Path to TLS certificate to use for public-inbox IMAP connections.";
+      };
+      key = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "/path/to/key.pem";
+        description = "Path to TLS key to use for public-inbox IMAP connections.";
+      };
+    };
+    nntp = {
+      enable = mkEnableOption "the public-inbox NNTP server";
+      port = mkOption {
+        type = with types; nullOr port;
+        default = 563;
+        description = ''
+          Listening port.
+          Set to null and use <code>systemd.sockets.public-inbox-nntpd.listenStreams</code>
+          if you need a more advanced listening.
+        '';
+      };
+      args = mkOption {
+        type = with types; listOf str;
+        default = ["-W0"];
+        description = "Command-line arguments to pass to ${manref "public-inbox-nntpd" 1}.";
+      };
+      cert = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "/path/to/fullchain.pem";
+        description = "Path to TLS certificate to use for public-inbox NNTP connections";
+      };
+      key = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "/path/to/key.pem";
+        description = "Path to TLS key to use for public-inbox NNTP connections.";
+      };
+    };
+    spamAssassinRules = mkOption {
+      type = with types; nullOr path;
+      default = "${cfg.package.sa_config}/user/.spamassassin/user_prefs";
+      description = "SpamAssassin configuration specific to public-inbox.";
+    };
+    settings = mkOption {
+      description = "Settings for the public-inbox config file.";
+      default = {};
+      type = types.submodule {
+        freeformType = gitIni.type;
+        options.publicinbox = mkOption {
+          default = {};
+          description = "public-inbox configuration.";
+          type = types.submodule {
+            freeformType = iniAttrs;
+            options.css = mkOption {
+              type = with types; listOf str;
+              default = [];
+              description = "The local path name of a CSS file for the PSGI web interface.";
+            };
```

Comment:
How about removing these `options.publicinbox = mkOption` and instead just declaring the suboptions directly:
```nix
    settings = mkOption {
      type = types.submodule {
        freeformType = gitIni.type;
        options.publicinbox.css = mkOption {
          type = with types; listOf str;
          default = [];
          description = "The local path name of a CSS file for the PSGI web interface.";
        };
```

Then you also shouldn't need to define the `iniAttrs` stuff above.

Reply:
Ah, right, I didn't know it was possible I've done the same for `options.publicinboxmda.*` and `options.publicinboxwatch.*`.

Reply:
@Infinisil, turns out I cannot declare the sub-options directly there because the parent option must have an`attrsOf` type for type-checking to work as expected. I've uploaded a corrected version.